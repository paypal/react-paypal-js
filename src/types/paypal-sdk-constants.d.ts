declare module "@paypal/sdk-constants/dist/module" {
    export const CURRENCY: Record<string, string>;
    export const FUNDING: Record<string, string>;
    export const JS_SDK_LIBRARIES: Record<string, string>;
    export const SDK_SETTINGS: Record<string, string>;
    export const SDK_QUERY_KEYS: Record<string, string>;
}
