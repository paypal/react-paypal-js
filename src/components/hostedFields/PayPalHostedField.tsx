import React from "react";
import type { FC } from "react";

import type { PayPalHostedFieldProps } from "../../types/payPalHostedFieldTypes";

/**
This `<PayPalHostedField />` component renders the [Braintree PayPal Hosted Fields](https://developer.paypal.com/braintree/docs/guides/hosted-fields/overview) for Braintree Merchants.
It relies on the `<PayPalHostedFieldsProvider />` parent component for managing state related to loading the JS SDK script
and execute some validations before the rendering the fields.

Use props for customizing your hosted fields. For example, here's how you would use the `style`, `className`, `id` options:

```jsx
    import {
        PayPalScriptProvider,
        PayPalHostedFieldsProvider,
        PayPalHostedField
    } from "@paypal/react-paypal-js";

    <PayPalScriptProvider options={{ "client-id": "test" }}>
        <PayPalHostedFieldsProvider
            createOrder={() => {
                // Manually call your server endpoint to create the client order
                return fetch("create_order_endpoint")
                    .then(response => response.json())
                    .then(order => order.id)
                    .catch(err => {
                        // Handle exceptions
                    })
            }}
        >
            <PayPalHostedField style={{ color: "red", border: "1px solid" }}
                id="card-number"
                hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.NUMBER}
                options={{ selector: "#card-number", placeholder: "4111 1111 1111 1111" }} />
            <PayPalHostedField id="cvv"
                hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.CVV}
                options={{ selector: "#cvv", placeholder: "CVV", maxlength: 3, maskInput: true }} />
            <PayPalHostedField id="expiration-date"
                hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.EXPIRATION_DATE}
                options={{ selector: "#expiration-date", placeholder: "MM/YY" }} />

        </PayPalHostedFieldsProvider>
    </PayPalScriptProvider>
```

To use the Braintree PayPal hosted fields you need to define at least three children:
    - A card number field
    - The CVV code from the client card
    - The expiration date

If the parent PayPalHostedFieldsProvider component doesn't found those three children will throw an Error

You can define the expiration date as a single field similar to the example above, 
or you are able to define it in two separate fields. One for the month and second for year.
You can delete the last children component form the example above and use these: 

```jsx
    <PayPalHostedField id="expiration-month"
        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.EXPIRATION_MONTH}
        options={{ selector: "#expiration-month", placeholder: "MM" }} />
    <PayPalHostedField id="expiration-year"
        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.EXPIRATION_YEAR}
        options={{ selector: "#expiration-year", placeholder: "YYYY" }} />
```

Note: Take care using multiple instances of the Braintree PayPal Hosted Fields.
In the case multiple instances are used in the same page with the same selector that will cause 
a failure in the render process.
*/
export const PayPalHostedField: FC<PayPalHostedFieldProps> = ({
    hostedFieldType, // eslint-disable-line @typescript-eslint/no-unused-vars
    options, // eslint-disable-line @typescript-eslint/no-unused-vars
    ...props
}) => <div {...props} />;
