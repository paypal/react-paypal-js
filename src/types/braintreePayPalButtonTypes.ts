import type { PayPalButtonsComponentProps } from "./paypalButtonTypes";
import type { BraintreeClient } from "./braintree/clientTypes";
import type {
    BraintreePayPalCheckout,
    BraintreePayPalCheckoutTokenizationOptions,
} from "./braintree/paypalCheckout";

export type BraintreeActions = {
    braintree: BraintreePayPalCheckout;
};

export type OnApproveBraintreeData = BraintreePayPalCheckoutTokenizationOptions;

export interface BraintreePayPalButtonsComponentProps
    extends Omit<
            PayPalButtonsComponentProps,
            "createOrder" | "onApprove" | "createBillingAgreement"
        >,
        Partial<
            Record<
                "createOrder" | "createBillingAgreement" | "onApprove",
                (
                    data: OnApproveBraintreeData | Record<string, unknown>,
                    actions: BraintreeActions
                ) => Promise<string | void>
            >
        > {}

export type BraintreeNamespace = {
    client: BraintreeClient;
    paypalCheckout: BraintreePayPalCheckout;
};
