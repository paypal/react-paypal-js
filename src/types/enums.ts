export enum SCRIPT_LOADING_STATE {
    INITIAL = "initial",
    PENDING = "pending",
    REJECTED = "rejected",
    RESOLVED = "resolved",
}

export enum SCRIPT_PROVIDER_DISPATCH_ACTION {
    LOADING_STATUS = "setLoadingStatus",
    RESET_OPTIONS = "resetOptions",
    SET_BRAINTREE_INSTANCE = "braintreeInstance",
}
