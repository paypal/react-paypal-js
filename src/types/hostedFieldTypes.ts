import type { PayPalHostedFieldsComponent } from "@paypal/paypal-js/types/components/hosted-fields";

export type HostedFieldsNamespace = {
    components: string | undefined;
    [DATA_NAMESPACE: string]: string | undefined;
};

export interface DecoratedPayPalHostedFieldsComponent
    extends PayPalHostedFieldsComponent {
    close(container: HTMLDivElement | null): void;
}

export interface HostedFieldProps {
    showLabel: boolean;
    label?: string;
}

export interface HostedFieldsComponentProps {
    showLabels: boolean;
    styles?: Record<string, unknown>;
    placeholder?: {
        number: string;
        cvv: string;
        expirationDate: string;
    };
    createOrder: () => Promise<string>;
}
