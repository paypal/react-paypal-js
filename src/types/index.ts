export * from "./braintreePayPalButtonTypes";
export * from "./enums";
export * from "./paypalButtonTypes";
export * from "./scriptProviderTypes";
export * from "./payPalHostedFieldTypes";
