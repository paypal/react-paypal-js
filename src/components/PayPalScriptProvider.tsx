import React, { FC, useEffect, useReducer } from "react";
import { loadScript } from "@paypal/paypal-js";

import {
    getScriptID,
    ScriptContext,
    scriptReducer,
} from "../context/scriptProviderContext";
import {
    SCRIPT_ID,
    DATA_SDK_INTEGRATION_SOURCE,
    DATA_SDK_INTEGRATION_SOURCE_VALUE,
    LOAD_SCRIPT_ERROR,
} from "../constants";
import type { ScriptProviderProps } from "../types";
import { SCRIPT_LOADING_STATE, DISPATCH_ACTION } from "../types";

/**
This `<PayPalScriptProvider />` component takes care of loading the JS SDK `<script>`.
It manages state for script loading so children components like `<PayPalButtons />` know when it's safe to use the `window.paypal` global namespace.

Note: You always should use this component as a wrapper for  `PayPalButtons`, `PayPalMarks`, `PayPalMessages` and `BraintreePayPalButtons` components.
 */
export const PayPalScriptProvider: FC<ScriptProviderProps> = ({
    options = { "client-id": "test" },
    children,
    deferLoading = false,
}: ScriptProviderProps) => {
    const [state, dispatch] = useReducer(scriptReducer, {
        options: {
            ...options,
            [SCRIPT_ID]: `${getScriptID(options)}`,
            [DATA_SDK_INTEGRATION_SOURCE]: DATA_SDK_INTEGRATION_SOURCE_VALUE,
        },
        loadingStatus: deferLoading
            ? SCRIPT_LOADING_STATE.INITIAL
            : SCRIPT_LOADING_STATE.PENDING,
    });

    useEffect(() => {
        if (
            deferLoading === false &&
            state.loadingStatus === SCRIPT_LOADING_STATE.INITIAL
        ) {
            return dispatch({
                type: DISPATCH_ACTION.LOADING_STATUS,
                value: SCRIPT_LOADING_STATE.PENDING,
            });
        }

        if (state.loadingStatus !== SCRIPT_LOADING_STATE.PENDING) return;

        let isSubscribed = true;

        loadScript(state.options)
            .then(() => {
                if (isSubscribed) {
                    dispatch({
                        type: DISPATCH_ACTION.LOADING_STATUS,
                        value: SCRIPT_LOADING_STATE.RESOLVED,
                    });
                }
            })
            .catch((err) => {
                console.error(`${LOAD_SCRIPT_ERROR} ${err}`);
                if (isSubscribed) {
                    dispatch({
                        type: DISPATCH_ACTION.LOADING_STATUS,
                        value: SCRIPT_LOADING_STATE.REJECTED,
                    });
                }
            });
        return () => {
            isSubscribed = false;
        };
    }, [state.options, deferLoading, state.loadingStatus]);

    return (
        <ScriptContext.Provider value={{ ...state, dispatch }}>
            {children}
        </ScriptContext.Provider>
    );
};
