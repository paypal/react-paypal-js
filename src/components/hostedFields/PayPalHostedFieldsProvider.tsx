import React, { useState, useEffect, useRef, Children } from "react";
import type { FC } from "react";

import { PayPalHostedFieldsContext } from "../../context/payPalHostedFieldsContext";
import { useScriptProviderContext } from "../../hooks/scriptProviderHooks";
import { DATA_NAMESPACE } from "../../constants";
import {
    generateHostedFieldsFromChildren,
    generateMissingHostedFieldsError,
} from "./utils";
import { validateHostedFieldChildren } from "./validators";
import { SCRIPT_LOADING_STATE } from "../../types/enums";
import { getPayPalWindowNamespace } from "../../utils";
import type { PayPalHostedFieldsComponentProps } from "../../types/payPalHostedFieldTypes";
import type {
    PayPalHostedFieldsComponent,
    HostedFieldsHandler,
} from "@paypal/paypal-js";

/**
This `<PayPalHostedFieldsProvider />` provider component wraps the form field elements and accepts props like `createOrder()`.

This provider component is designed to be used with the `<PayPalHostedField />` component.

Warning: If you don't see anything in the screen probably your client is ineligible.
To handle this problem make sure to use the prop `notEligibleError` and pass a component with a custom message.
Take a look to this link if that is the case: https://developer.paypal.com/docs/checkout/advanced/integrate/
*/
export const PayPalHostedFieldsProvider: FC<
    PayPalHostedFieldsComponentProps
> = ({ styles, createOrder, notEligibleError, children }) => {
    const childrenList = Children.toArray(children);
    const [{ options, loadingStatus }] = useScriptProviderContext();
    const [isEligible, setIsEligible] = useState(true);
    const [cardFields, setCardFields] = useState<HostedFieldsHandler | null>(
        null
    );
    const hostedFieldsContainerRef = useRef<HTMLDivElement>(null);
    const hostedFields = useRef<PayPalHostedFieldsComponent>();
    const [, setErrorState] = useState(null);

    /**
     * Executed on the mount process to validate the children
     */
    useEffect(() => {
        validateHostedFieldChildren(childrenList);
    }, []); // eslint-disable-line react-hooks/exhaustive-deps

    useEffect(() => {
        // Only render the hosted fields when script is loaded and hostedFields is eligible
        if (!(loadingStatus === SCRIPT_LOADING_STATE.RESOLVED)) {
            return;
        }
        // Get the hosted fields from the [window.paypal.HostedFields] SDK
        hostedFields.current = getPayPalWindowNamespace(
            options[DATA_NAMESPACE]
        ).HostedFields;

        if (!hostedFields.current) {
            throw new Error(
                generateMissingHostedFieldsError({
                    components: options.components,
                    [DATA_NAMESPACE]: options[DATA_NAMESPACE],
                })
            );
        }
        if (!hostedFields.current.isEligible()) {
            return setIsEligible(false);
        }
        // Clean all the fields before the rerender
        if (cardFields) {
            cardFields.teardown();
        }

        hostedFields.current
            .render({
                // Call your server to set up the transaction
                createOrder: createOrder,
                styles: styles,
                fields: generateHostedFieldsFromChildren(childrenList),
            })
            .then((cardFieldsInstance) => {
                setCardFields(cardFieldsInstance);
            })
            .catch((err) => {
                setErrorState(() => {
                    throw new Error(
                        `Failed to render <PayPalHostedFieldsProvider /> component. ${err}`
                    );
                });
            });
    }, [loadingStatus, styles]); // eslint-disable-line react-hooks/exhaustive-deps

    return (
        <div ref={hostedFieldsContainerRef}>
            {isEligible ? (
                <PayPalHostedFieldsContext.Provider value={cardFields}>
                    {children}
                </PayPalHostedFieldsContext.Provider>
            ) : (
                notEligibleError
            )}
        </div>
    );
};
