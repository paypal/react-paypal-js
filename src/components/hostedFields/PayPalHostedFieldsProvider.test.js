import React from "react";
import { render, waitFor } from "@testing-library/react";
import { ErrorBoundary } from "react-error-boundary";
import { loadScript } from "@paypal/paypal-js";

import { PayPalScriptProvider } from "../PayPalScriptProvider";
import { PayPalHostedFieldsProvider } from "./PayPalHostedFieldsProvider";
import { PayPalHostedField } from "./PayPalHostedField";
import { PAYPAL_HOSTED_FIELDS_TYPES } from "../../types/enums";

const onError = jest.fn();
const wrapper = ({ children }) => (
    <ErrorBoundary fallback={<div>Error</div>} onError={onError}>
        {children}
    </ErrorBoundary>
);

jest.mock("@paypal/paypal-js", () => ({
    loadScript: jest.fn(),
}));

describe("PayPalHostedFieldsProvider", () => {
    const isEligible = jest.fn();

    beforeEach(() => {
        document.body.innerHTML = "";

        window.paypal = {
            HostedFields: {
                isEligible: isEligible.mockReturnValue(true),
                render: jest.fn().mockResolvedValue({
                    teardown: jest.fn(),
                }),
            },
        };

        loadScript.mockResolvedValue(window.paypal);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    test("should throw an Error using the component without the PayPalScriptProvider", () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();

        render(<PayPalHostedFieldsProvider />, { wrapper });
        expect(onError.mock.calls[0][0].message).toEqual(
            "usePayPalScriptReducer must be used within a PayPalScriptProvider"
        );
        spyConsoleError.mockRestore();
    });

    test("should throw an Error using the component with PayPalScriptProvider without data-client-token", () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();

        render(
            <PayPalScriptProvider>
                <PayPalHostedFieldsProvider />
            </PayPalScriptProvider>,
            { wrapper }
        );
        expect(onError.mock.calls[0][0].message).toEqual(
            "A client token wasn't found in the provider parent component"
        );
        spyConsoleError.mockRestore();
    });

    test("should throw an Error using the component without children", () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();

        render(
            <PayPalScriptProvider
                options={{
                    "client-id": "test-client",
                    currency: "USD",
                    intent: "authorize",
                    components: "hosted-fields",
                    "data-client-token": "test-data-client-token",
                }}
            >
                <PayPalHostedFieldsProvider />
            </PayPalScriptProvider>,
            { wrapper }
        );
        expect(onError.mock.calls[0][0].message).toEqual(
            "To use HostedFields you must use it with at least 3 children with types: [number, cvv, expirationDate] includes"
        );
        spyConsoleError.mockRestore();
    });

    test("should throw an Error about missing HostedFields in paypal SDK because hosted-fields isn't imported in components", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        window.paypal = {};

        render(
            <PayPalScriptProvider
                options={{
                    "client-id": "test-client",
                    currency: "USD",
                    intent: "authorize",
                    "data-client-token": "test-data-client-token",
                }}
            >
                <PayPalHostedFieldsProvider>
                    <PayPalHostedField
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.NUMBER}
                        options={{ selector: "number" }}
                    />
                    <PayPalHostedField
                        hostedFieldType={
                            PAYPAL_HOSTED_FIELDS_TYPES.EXPIRATION_DATE
                        }
                        options={{ selector: "expiration" }}
                    />
                    <PayPalHostedField
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.CVV}
                        options={{ selector: "cvv" }}
                    />
                </PayPalHostedFieldsProvider>
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => {
            expect(onError.mock.calls[0][0].message).toEqual(
                "Unable to render <PayPalHostedFieldsProvider /> because window.paypal.HostedFields is undefined." +
                    "\nTo fix the issue, add 'hosted-fields' to the list of components passed to the parent PayPalScriptProvider: " +
                    "<PayPalScriptProvider options={{ components: 'hosted-fields'}}>"
            );
        });
        spyConsoleError.mockRestore();
    });

    test("should return immediately when script provider is rejected", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        loadScript.mockRejectedValue(new Error("Unknown error"));

        render(
            <PayPalScriptProvider
                options={{
                    "client-id": "test-client",
                    currency: "USD",
                    intent: "authorize",
                    components: "hosted-fields",
                    "data-client-token": "test-data-client-token",
                }}
            >
                <PayPalHostedFieldsProvider>
                    <PayPalHostedField
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.NUMBER}
                        options={{ selector: "number" }}
                    />
                    <PayPalHostedField
                        hostedFieldType={
                            PAYPAL_HOSTED_FIELDS_TYPES.EXPIRATION_DATE
                        }
                        options={{ selector: "expiration" }}
                    />
                    <PayPalHostedField
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.CVV}
                        options={{ selector: "cvv" }}
                    />
                </PayPalHostedFieldsProvider>
            </PayPalScriptProvider>
        );
        await waitFor(() => {
            expect(loadScript).toBeCalled();
        });
        spyConsoleError.mockRestore();
    });

    test("should remove hostedfields components when unilegible", async () => {
        isEligible.mockReturnValue(false);

        const { container } = render(
            <PayPalScriptProvider
                options={{
                    "client-id": "test-client",
                    currency: "USD",
                    intent: "authorize",
                    components: "hosted-fields",
                    "data-client-token": "test-data-client-token",
                }}
            >
                <PayPalHostedFieldsProvider>
                    <PayPalHostedField
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.NUMBER}
                        options={{ selector: "number" }}
                    />
                    <PayPalHostedField
                        hostedFieldType={
                            PAYPAL_HOSTED_FIELDS_TYPES.EXPIRATION_DATE
                        }
                        options={{ selector: "expiration" }}
                    />
                    <PayPalHostedField
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.CVV}
                        options={{ selector: "cvv" }}
                    />
                </PayPalHostedFieldsProvider>
            </PayPalScriptProvider>
        );
        await waitFor(() => {
            expect(container.querySelector(".number")).toEqual(null);
        });
        expect(container.querySelector(".expiration")).toEqual(null);
        expect(container.querySelector(".cvv")).toEqual(null);
    });

    test("should throw an Error on hosted fields render process exception", async () => {
        const spyConsoleError = jest
            .spyOn(console, "error")
            .mockImplementation();
        window.paypal.HostedFields.render = jest
            .fn()
            .mockRejectedValue(new Error("Failing rendering hostedFields"));

        render(
            <PayPalScriptProvider
                options={{
                    "client-id": "test-client",
                    currency: "USD",
                    intent: "authorize",
                    "data-client-token": "test-data-client-token",
                }}
            >
                <PayPalHostedFieldsProvider>
                    <PayPalHostedField
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.NUMBER}
                        options={{ selector: "number" }}
                    />
                    <PayPalHostedField
                        hostedFieldType={
                            PAYPAL_HOSTED_FIELDS_TYPES.EXPIRATION_DATE
                        }
                        options={{ selector: "expiration" }}
                    />
                    <PayPalHostedField
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.CVV}
                        options={{ selector: "cvv" }}
                    />
                </PayPalHostedFieldsProvider>
            </PayPalScriptProvider>,
            { wrapper }
        );

        await waitFor(() => {
            expect(onError.mock.calls[0][0].message).toEqual(
                "Failed to render <PayPalHostedFieldsProvider /> component. Error: Failing rendering hostedFields"
            );
        });
        spyConsoleError.mockRestore();
    });

    test("should render hosted fields", async () => {
        const { container, rerender } = render(
            <PayPalScriptProvider
                options={{
                    "client-id": "test-client",
                    currency: "USD",
                    intent: "authorize",
                    "data-client-token": "test-data-client-token",
                    components: "hosted-fields",
                }}
            >
                <PayPalHostedFieldsProvider>
                    <PayPalHostedField
                        className="number"
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.NUMBER}
                        options={{ selector: ".number" }}
                    />
                    <PayPalHostedField
                        className="expiration"
                        hostedFieldType={
                            PAYPAL_HOSTED_FIELDS_TYPES.EXPIRATION_DATE
                        }
                        options={{ selector: ".expiration" }}
                    />
                    <PayPalHostedField
                        className="cvv"
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.CVV}
                        options={{ selector: ".cvv" }}
                    />
                </PayPalHostedFieldsProvider>
            </PayPalScriptProvider>
        );

        await waitFor(() => {
            expect(window.paypal.HostedFields.render).toBeCalled();
        });
        expect(
            container.querySelector(".number") instanceof HTMLDivElement
        ).toBeTruthy();
        expect(
            container.querySelector(".expiration") instanceof HTMLDivElement
        ).toBeTruthy();
        expect(container.querySelector(".cvv")).toBeTruthy();

        // Rerender the component with new styles props
        // Shouldn't change the hostedFields refs when rerendering
        rerender(
            <PayPalScriptProvider
                options={{
                    "client-id": "test-client",
                    currency: "USD",
                    intent: "authorize",
                    "data-client-token": "test-data-client-token",
                }}
            >
                <PayPalHostedFieldsProvider
                    styles={{
                        ".valid": { color: "#28a745" },
                        ".invalid": { color: "#dc3545" },
                    }}
                >
                    <PayPalHostedField
                        className="number"
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.NUMBER}
                        options={{ selector: ".number" }}
                    />
                    <PayPalHostedField
                        className="expiration"
                        hostedFieldType={
                            PAYPAL_HOSTED_FIELDS_TYPES.EXPIRATION_DATE
                        }
                        options={{ selector: ".expiration" }}
                    />
                    <PayPalHostedField
                        className="cvv"
                        hostedFieldType={PAYPAL_HOSTED_FIELDS_TYPES.CVV}
                        options={{ selector: ".cvv" }}
                    />
                </PayPalHostedFieldsProvider>
            </PayPalScriptProvider>
        );
        await waitFor(() => {
            expect(window.paypal.HostedFields.render).toBeCalledTimes(2);
        });
    });
});
