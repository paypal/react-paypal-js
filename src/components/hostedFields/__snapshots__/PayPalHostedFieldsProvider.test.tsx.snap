// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PayPalHostedFieldsProvider should throw an Error using the component without the PayPalScriptProvider 1`] = `"usePayPalScriptReducer must be used within a PayPalScriptProvider"`;
